clang/utils/TableGen/ClangProgModelBuiltinEmitter.cpp

```diff
@@ -848,6 +922,228 @@ void BuiltinNameEmitter::EmitQualTypeFinder() {
+  Size = T->getValueAsInt("ElementSize");
+  VecWidth = T->getValueAsInt("VecWidth");
+  IsConst = T->getValueAsBit("IsConst");
+  IsVolatile = T->getValueAsBit("IsVolatile");
+  IsPointer = T->getValueAsBit("IsPointer");
+  IsBuiltin =
+      !T->isSubClassOf("FundamentalType") && !T->isSubClassOf("VectorType");
+
+  AddrSpace = ToOCLAddressSpace(T->getValueAsString("AddrSpace"));
+}
+
+std::string JSONBuiltinInterfaceEmitter::TypeDesc::GetBaseTypeAsStr() const {
+  if (Name == "void") {
+    return Name.str();
+  }
+
+  std::string TypeStr = TypePrefix;
+  if (IsBuiltin) {
+    return (TypeStr + "_" + Name).str();
+  }
+  if (VecWidth != 1) {
+    TypeStr += "_vec" + std::to_string(VecWidth);
+  }
+  if (IsInteger) {
+    if (Size == 1)
+      TypeStr += "_bool";
+    else {
+      TypeStr += IsSigned ? "_int" : "_uint";
+      TypeStr += std::to_string(Size);
+    }
+  } else {
+    if (IsFloat) {
+      TypeStr += "_fp";
+      TypeStr += std::to_string(Size);
+    } else {
+      llvm_unreachable("Unknown type");
+    }
+  }
+  TypeStr += "_t";
+
+  return TypeStr;
+}
+
+void JSONBuiltinInterfaceEmitter::Emit() {
+
+  // Parse the Records to populate the internal lists.
+  GetOverloads();
+  GroupBySignature();
+  ExpandTypes();
+
+  EmitBuiltins();
+}
+
+// Generate all possible types represented by the input record.
+void JSONBuiltinInterfaceEmitter::ExpandType(const Record *Ty) {
+  auto ExpansionIt = ExpandedTypes.find(Ty);
+
+  if (ExpansionIt != ExpandedTypes.end())
+    return;
+  llvm::SmallVector<std::string, 11> &Expansion = ExpandedTypes[Ty];
+
+  JSONBuiltinInterfaceEmitter::TypeDesc TypeDesc(Ty);
+
+  bool IsCompound = Ty->isSubClassOf("CompoundType");
+  if (IsCompound) {
+    auto *EltRecord = Ty->getValueAsDef("ElementType");
+    ExpandType(EltRecord);
+    Expansion = ExpandedTypes[EltRecord];
+    for (auto &TypeStr : Expansion) {
+      if (TypeDesc.IsPointer) {
+        if (TypeDesc.IsConst)
+          TypeStr += " const";
+        if (TypeDesc.IsVolatile)
+          TypeStr += " volatile";
+        TypeStr += " ";
+        TypeStr += TypeDesc.AddrSpace;
+        TypeStr += " *";
+      }
+    }
+
+    return;
+  }
+
+  if (Ty->isSubClassOf("GenericType")) {
+    assert(!Expansion.size() && "Types already expanded");
+    auto NVec = Ty->getValueAsDef("VectorList")->getValueAsListOfInts("List");
+    auto NTypes = Ty->getValueAsDef("TypeList")->getValueAsListOfDefs("List");
+    for (auto SubT : NTypes) {
+      for (auto VecWidth : NVec) {
+        JSONBuiltinInterfaceEmitter::TypeDesc TypeDesc(SubT);
+        TypeDesc.VecWidth = VecWidth;
+        Expansion.emplace_back(TypeDesc.GetBaseTypeAsStr());
+      }
+    }
+
+    return;
+  }
+
+  Expansion.emplace_back(TypeDesc.GetBaseTypeAsStr());
+}
+
+void JSONBuiltinInterfaceEmitter::ExpandTypes() {
+  for (auto T : Records.getAllDerivedDefinitions("Type")) {
+    ExpandType(T);
+  }
+}
+
+void JSONBuiltinInterfaceEmitter::EmitBuiltins() {
+  struct FnDesc {
```

Comment:
So why can't we just emit as we go like the other emitters?  Saving all the strings until the end seems like a waste.

Reply:
Overloads are sorted by signature where I want them sorted by names.

Reply:
Where does this preference come from?  Is there good reason to alter the order and complicate the code so much further?

Reply:
> Where does this preference come from?

Generating a correct output.

The function name is used as key, I would loose information if I didn't.